keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1080 @@
+```
+const GroupName = "counter.dra.config.k8s.io"
+const KindName = "Capacity"
+const Version = "v1beta1"
+
+type Capacity struct {
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	metav1.TypeMeta `json:",inline"`
+
+	Count int64 `json:"count"`
+}
+```
+
+The kubelet then combines all of this information in a new NodeResourceCapacity
+object which contains the JSON representation of a capacity type in a
+`runtime.RawExtension`. The kubelet itself never has to interpret that data.
+
+### Using numeric parameters as claim parameters
+
+The numeric model also defines a type for claim parameters. In the simple example
+above this is the same `Capacity` type, but more complex models will have
+a more complex capacity type and a simpler parameter type.
+
+When defining the CRD for the claim parameters, the driver developer replicates
+those fields from the numeric model which are relevant for the driver and adds
+custom validation. For a complex model, this can be a true subset of the fields
+in the parameter type. The validation can set bounds that are specific to the
+driver. One can think of this as "duck typing": the CRD doesn't need to include
+the numeric type, a subset of it just needs to behave like one.
+
+To enable extracting the numeric parameters, they must be located at a specific
+JSON path (e.g. "spec"). "Kind" and "apiVersion" must be included at that
+path. Users don't need to set those, the CRD can add them via defaults. This is
+best explained with an example:
+
+```
+const GroupName = "dra.e2e.example.com"
+const KindName = "ClaimParameter"
+const Version = "v1alpha1"
+
+type ClaimParameter struct {
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	metav1.TypeMeta   `json:",inline"`
+
+	Spec ClaimParameterSpec `json:"spec"`
+}
+
+type ClaimParameterSpec struct {
+	TypeMeta `json:",inline"` // Not metav1.TypeMeta! See below.
+
+	// +kubebuilder:default=1
+	// +kubebuilder:validation:Minimum=1
+    // +kubebuilder:validation:Maximum=16
+	Count int64 `json:"count,omitempty"`
+
+	Env map[string]string `json:"env,omitempty"`
+}
+
+type TypeMeta struct {
+	// +kubebuilder:default=Capacity
+	// +kubebuilder:validation:Enum=Capacity
+	Kind string `json:"kind,omitempty"`
+
+	// +kubebuilder:default=counter.dra.config.k8s.io/v1alpha1
+	// +kubebuilder:validation:Enum=counter.dra.config.k8s.io/v1alpha1
+	APIVersion string `json:"apiVersion,omitempty"`
+}
+```
+
+In this example, the driver supports the `Count` numeric parameter and adds
+environment variables as configuration parameters. There is a 1:1 relationship
+between the CRD type and the numeric type that it is using. Users define
+parameters like they would without numeric parameters:
+
+```
+kind: ClaimParameter
+apiVersion: dra.e2e.example.com
+spec:
+  count: 10
+  env:
+    HELLO_WORLD: "Thanks for the fish."
+```
+
+### Notes/Constraints/Caveats
+
+For this proposal to work, kube-scheduler must have permission to read, list and
```

Comment:
I hadn't thought about that possibility. This should indeed work.

But from a practical perspective it raises a challenge: does the scheduler have a well-defined service account name? Otherwise vendor's won't know what to put into their role binding. We probably have to warn vendors that they need to be flexible here in their driver deployment.

Reply:
I poked around a bit. There is a default bootstrap cluster role binding for the `system:kube-scheduler` user:

```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  annotations:
    rbac.authorization.kubernetes.io/autoupdate: "true"
  creationTimestamp: "2023-12-19T01:18:51Z"
  labels:
    kubernetes.io/bootstrapping: rbac-defaults
  name: system:kube-scheduler
  resourceVersion: "141"
  uid: 0c0ec4fa-27ea-4aae-99dc-8069e8c80334
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: system:kube-scheduler
- apiGroup: rbac.authorization.k8s.io
  kind: User
  name: system:kube-scheduler
```

That is a well-known user for kube-scheduler, so the recommendation for driver authors would be to create a cluster role that provides access to their CRs, and add a cluster role binding for that user. If some distribution of K8s does things differently, they would need to modify the driver installation.

Reply:
https://kubernetes.io/docs/reference/access-authn-authz/rbac/#core-component-roles